### ISSUE TITLE
How to specify the Output with model with Ragged Input

### ISSUE THREAD

**Issue by alexey-laletin-singularitynet** [triton-inference-server/server#8261]
**Is your feature request related to a problem? Please describe.**
I work with Audio Models which take **[channels, length]** Tensors as an Input and returns an Output with same format.

Let's assume that we have 3 audio tensors: _[1, 1, 1]_, _[2, 2]_, _[3, 3, 3, 3]_. (Let's skip the details about sample rate etc.)
So I cannot use dynamic batching, because the shapes are not the same (_[1, 3]_, _[1, 2]_, _[1, 4]_). 

Thus I can use [Ragged Input](https://docs.nvidia.com/deeplearning/triton-inference-server/user-guide/docs/user_guide/ragged_batching.html) to concatinate inputs into one and split it afterwards. Following the example we would get this:

```
INPUT: [1, 1, 1, 2, 2, 3, 3, 3, 3]
BATCHED_INPUT: [[1, 3], [1, 2], [1, 3]]
```

Then I preprocess this input by spliting and padding parts up to max length, so my model can process this at once:

```
                               [1, 1, 1]       [1, 1, 1, 0]
[1, 1, 1, 2, 2, 3, 3, 3, 3] -> [2, 2]       -> [2, 2, 0, 0]
                               [3, 3, 3, 3]    [3, 3, 3, 3]
```

Then I run my model and get results. Let's also assume that output's shape differs from input's one.

```
[1, 1, 1, 0]                                [4, 4, 4, 4, 0]
[2, 2, 0, 0] -> | AUDIO_TO_AUDIO_MODEL | -> [5, 5, 5, 0, 0]
[3, 3, 3, 3]                                [6, 6, 6, 6, 6]
```

**And here is the problem with which I struggle comes in.** I need to decouple my results to users. But there is no documented way to return this result in the way simmular to Ragged Input. For example:

```
[4, 4, 4, 4, 0]                          [4, 4, 4, 4]
[5. 5. 5. 0. 0] -> | REMOVE_SILENCE | -> [5, 5, 5]       -> OUTPUT:         [4, 4, 4, 4, 5, 5, 5, 6, 6, 6, 6, 6]
[6, 6, 6, 6, 6]                          [6, 6, 6, 6, 6]    BATCHED_OUTPUT: [[1, 4], [1, 3], [1, 5]]
```

But I cannot find any way to implement this behavior. I found that there is [Batched Output in Protobuf config](https://github.com/triton-inference-server/common/blob/main/protobuf/model_config.proto#L591), but not found any examples.

**Describe the solution you'd like**
From earlier description I'd like to specify the Ragged Output in config.pbtxt like this:

```
...
input [
  {
    name: "input_audio"
    data_type: TYPE_FP64
    dims: [1, -1]
  }
]
batch_input [
  {
    kind: BATCH_ITEM_SHAPE 
    target_name: "input_audio_shapes"
    data_type: TYPE_INT32
    source_input: "input_audio"
  }
]

output [
  {
    name: "output_audio"
    data_type: TYPE_FP64
    dims: [1, -1]
  }
]
batch_output [
  {
    kind: BATCH_ITEM_SHAPE 
    target_name: "output_name_shapes"
    data_type: TYPE_INT32
    source_input: "input_audio" # MAY BE NEEDED TO SEND THE RESULT TO CORRECT USER
  }
]
...
```

So, whole pipeline looks like this (Sry. for long ASCII, I believe that it's demonstrative):

```
USER_1: [1, 1, 1]    \   | python_backend                                                                                                                                               |  /> USER_1: [4, 4, 4, 4]
                      \  |                                   [1, 1, 1]       [1, 1, 1, 0]    |       |    [4, 4, 4, 4, 0]    [4, 4, 4, 4]                                               | /
USER_2: [2, 2]       --> | I: [1, 1, 1, 2, 2, 3, 3, 3, 3] -> [2, 2]       -> [2, 2, 0, 0] -> | MODEL | -> [5, 5, 5, 0, 0] -> [5, 5, 5]       -> O: [4, 4, 4, 4, 5, 5, 5, 6, 6, 6, 6, 6] | --> USER_2: [5, 5, 5]
                      /  | BATCHED_I [[1,3],[1,2],[1,4]]     [3, 3, 3, 3]    [3, 3, 3, 3]    |       |    [6, 6, 6, 6, 6]    [6, 6, 6, 6, 6]    BATCHED_O: [[1,4],[1,3],[1,5]]          | \
USER_3: [3, 3, 3, 3] /   |                                                                                                                                                              |  \> USER_3: [6, 6, 6, 6, 6]
                                                                                                                            ^----------------------------------------------------------------------------------------^
                                                                                                                                        I ask to add feature to be able to configurate this behaviour
```

**Describe alternatives you've considered**
If any others feature in Triton allow to implement the behaviour I look for - it would be enougth. Some custom Ensemble Schedulers and Guides to implement them or showing how to trace Request User ID from a start of pipeline to an end, so I can decouple bathed model's output myself in Python Backend.

**Additional context**
Here I assume that all inputs have first dimention as [1] (so, in terms of audio, all of them are mono audio). Thus I don't think about situations where Inputs have different channel's dimention. Without it the problem becomes much more complex than one I can solve.

**Comment by alexey-laletin-singularitynet** [comment]
I figured it out.

So, If you want to do the same pipeline as I mentioned earlier, you can do this in Python Backend.

First of all, specify `allow_ragged_batch` and `dynamic_batching` in `config.pbtxt` of the model you want to add Ragged Input into.

For example:
```
name: "your_model"
backend: "python"

max_batch_size: 16

dynamic_batching {
    preferred_batch_size: [ 2, 4, 8, 16 ]
}

input [
  {
    name: "input_audio"
    data_type: TYPE_FP64
    dims: [1, -1]
    allow_ragged_batch: true
  }
]

output [
    {
        name: "output_audio"
        data_type: TYPE_FP64
        dims: [ 1, -1 ]
    }
]
```

This config allows you to get more than one request as a `requests` input in python backend. 
Next step is to write python backend. For example:

```python
...
class TritonPythonModel:
    def initialize(self, args):
        self._model = SOME_YOUR_MODEL_INITIALIZATION

    def execute(self, requests):
        responses = []

        logger.info(f"Got {len(requests)} requests")

        input_tensors: list[torch.Tensor] = []
        input_tensors_lengths: list[int] = []

        # Because of Ragged Input you can get len(requests) >= 1, otherwise len(requests) is always 1.
        for request in requests:
            input_tensor = torch.Tensor(
                pb_utils.get_input_tensor_by_name(request, "input_audio").as_numpy().squeeze(axis=0)
            )

            input_tensors.append(input_tensor)
            input_tensors_lengths.append(input_tensor.shape[1])

        padded_input_tensors = self._pad_tensors(input_tensors, input_tensors_lengths)

        padded_output_tensors = self._model(padded_origin_tensors)

        output_tensors = self._unpad_tensors(padded_output_tensors, input_tensors_lengths)

        for output_tensor in output_tensors:
            inference_response = pb_utils.InferenceResponse(
                output_tensors=[
                    pb_utils.Tensor("output_audio", output_tensor.numpy())
                ]
            )
            responses.append(inference_response)

        return responses

    def _pad_tensors(self, tensors: list[torch.Tensor], lengths: list[int]) -> torch.Tensor:
        max_length = np.max(lengths)
        result = torch.cat(
            [pad(tensor, (0, max_length - length), "constant", 0) for tensor, length in zip(tensors, lengths)],
            dim=0,
        )
        return result

    def _unpad_tensors(self, tensors: torch.Tensor, lengths: list[int]) -> list[torch.Tensor]:
        return [tensors[i, :length].reshape(1, -1) for i, length in enumerate(lengths)]
```

And now model is able to get several requests and pad them into one matrix. Then you need to return the same number of responses in original ordering. Profit.
